snuba/datasets/cdc/__init__.py

```diff
@@ -1,14 +1,75 @@
+from typing import Any, Optional, MutableMapping, Sequence, Type
+from snuba.clusters.cluster import get_cluster
+from snuba.clusters.storage_sets import StorageSetKey
+from snuba.datasets.dataset_schemas import StorageSchemas
 from snuba.datasets.storage import WritableTableStorage
+from snuba.datasets.storages import StorageKey
+from snuba.datasets.table_storage import KafkaStreamLoader, TableWriter
+from snuba.datasets.cdc.cdcprocessors import CdcMessageRow
+from snuba.query.processors import QueryProcessor
+from snuba.replacers.replacer_processor import ReplacerProcessor
+from snuba.snapshots.loaders.single_table import SingleTableBulkLoader
 
 
-class CdcStorage(WritableTableStorage):
+class CdcTableWriter(TableWriter):
     def __init__(
-        self, *, default_control_topic: str, postgres_table: str, **kwargs,
+        self, postgres_table: str, message_row: Type[CdcMessageRow], **kwargs,
     ):
         super().__init__(**kwargs)
+        self.__postgres_table = postgres_table
+        self.__message_row = message_row
+
+    def get_bulk_loader(self, source, dest_table) -> SingleTableBulkLoader:
+        return SingleTableBulkLoader(
+            source=source,
+            source_table=self.__postgres_table,
+            dest_table=dest_table,
+            row_processor=lambda row: self.__message_row.from_bulk(row).to_clickhouse(),
+        )
+
+
+class CdcStorage(WritableTableStorage):
+    def __init__(
+        self,
+        *,
+        default_control_topic: str,
+        postgres_table: str,
+        message_row: Type[CdcMessageRow],
+        storage_key: StorageKey,
+        storage_set_key: StorageSetKey,
+        schemas: StorageSchemas,
+        query_processors: Sequence[QueryProcessor],
+        stream_loader: KafkaStreamLoader,
+        replacer_processor: Optional[ReplacerProcessor] = None,
+        writer_options: Optional[MutableMapping[str, Any]] = None,
+    ):
+        super().__init__(
+            storage_key=storage_key,
+            storage_set_key=storage_set_key,
+            schemas=schemas,
+            query_processors=query_processors,
+            stream_loader=stream_loader,
+            replacer_processor=replacer_processor,
+            writer_options=writer_options,
+        )
         self.__default_control_topic = default_control_topic
         self.__postgres_table = postgres_table
 
+        write_schema = schemas.get_write_schema()
+        assert write_schema is not None
+        self.__table_writer = CdcTableWriter(
+            postgres_table=postgres_table,
+            message_row=message_row,
+            cluster=get_cluster(storage_set_key),
+            write_schema=write_schema,
+            stream_loader=stream_loader,
+            replacer_processor=replacer_processor,
+            writer_options=writer_options,
+        )
```

Comment:
I think this is a smell. This way you are instantiating two table writer for this class. One in the parent implementation that returns, which is a TableWriter and one in this class that overrides the instance above which is a CdcTableWriter. This basically breaks the encapsulation of a private method of the parent class and can easily break if someone adds a public method in the parent class that depends on the private __table_writer. They would never find out that they are acting or returning the wrong object. So it is risky.

I think this was solved before by passing the TableWriter to the constructor. If we need to change that behavior, then some additional refactoring is needed. Ideally there should not be two different instances (of two different classes) of the TableWriter around, that should be the sign that something is wrong.

Reply:
Yeah, deleted it now.